pkg/queue/workers/retention.go

```diff
@@ -0,0 +1,204 @@
+		// we check for errSerializingHearbeat so we don't cause
+		// a recursion call
+		if err == nil || err == errSerializingHearbeat {
+			return
+		}
+		message := err.Error()
+		progress.ErrorMessage = &message
+		_ = sendRetentionProgress(progress, heartbeats)
+	}()
+
+	spec := retentionSpec{}
+	err = json.Unmarshal(task.Spec, &spec)
+	if err != nil {
+		return fmt.Errorf("can not parse the retention task spec %w", err)
+	}
+
+	span.SetTag("spec.queue", spec.QueueName)
+	span.SetTag("spec.taskType", spec.TaskType)
+	span.SetTag("spec.status", spec.Status)
+	span.SetTag("spec.age", spec.Age.String())
+
+	// initial heartbeat
+	err = sendRetentionProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+	result, err := h.db.ExecContext(ctx, spec.SQL)
+	duration := time.Since(now).Milliseconds()
+	progress.Duration = &duration
+
+	if err != nil {
+		return fmt.Errorf("retention execution failed: %w", err)
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("can not get rows affected: %w", err)
+	}
+
+	progress.RowsAffected = &rowsAffected
+	err = sendRetentionProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	log.WithField("rowCount", rowsAffected).Debug("retention finished")
+	return nil
+}
+
+func sendRetentionProgress(progress retentionProgress, heartbeats chan<- queue.Progress) (err error) {
+	logrus.
+		WithField("method", "sendRetentionProgress").
+		Debugf("%+v", progress)
+
+	bytes, err := json.Marshal(progress)
+	if err != nil {
+		logrus.Error(err)
+		return errSerializingHearbeat
+	}
+
+	heartbeats <- bytes
+	return nil
+}
+
+// AssertRetentionSchedule creates a new queue retention tasks for the supplied queue, finished tasks matching
+// the supplied parameters will be deleted
+func AssertRetentionSchedule(ctx context.Context, scheduler queue.Scheduler, queueName string, taskType queue.TaskType, status queue.TaskStatus, filter squirrel.Sqlizer, age time.Duration) error {
+	spec := createRetentionSpec(queueName, taskType, status, filter, age)
+	specBytes, err := json.Marshal(spec)
+	if err != nil {
+		return fmt.Errorf("can not build retention task spec: %w", err)
+	}
+	// randomly distribute the retention tasks throughout the hour
+	when := rand.Intn(60)
+	retentionSchedule := queue.TaskScheduleRequest{
+		TaskBase: queue.TaskBase{
+			Queue: MaintenanceTaskQueue,
+			Type:  RetentionTask,
+			Spec:  specBytes,
+		},
+		CronSchedule: fmt.Sprintf("%d * * * *", when), // every hour at minute "when"
+	}
+
+	return scheduler.AssertSchedule(ctx, retentionSchedule)
+}
+
+//createRetentionSpec builds the task retention job spec. It is split out to simplify test setup
+func createRetentionSpec(queueName string, taskType queue.TaskType, status queue.TaskStatus, filter squirrel.Sqlizer, age time.Duration) retentionSpec {
+	spec := retentionSpec{
+		QueueName: queueName,
+		TaskType:  taskType,
+		Status:    status,
+		Age:       age,
+		SQL:       "",
```

Comment:
I think we should not store the rest of the parameters in the spec, reading the `SQL` should be informative enough from the debugging perspective and the rest of the spec fields are used only to generate the SQL statement.

I think what we should do is to define a generic `SQLExecTask` that executes an arbitrary SQL statement. Then make `CreateTaskRetentionSpec` public which would return `SQLExecTaskSpec`.

The rest of the implementation should be in the service (basically asserting the schedule).

Reply:
yes, i like splitting out a "SQLExecTask"